### PR TITLE
ITE driver/watchdog: add watchdog timer for it8xxx2

### PR DESCRIPTION
**Pull request by RuibinChang**
Add watchdog timer for it8xxx2.

Signed-off-by: Ruibin Chang <ruibin.chang@ite.com.tw>

### CODE REVIEW

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 1]
```diff
@@ -1116,30 +1116,75 @@
...
+	volatile uint8_t reserved6[25];
+	/* 0x060: External WDT Counter Observation Low Byte */
+	volatile uint8_t EWDCNTOLR;
+	/* 0x061: External WDT Counter Observation High Byte */
+	volatile uint8_t EWDCNTOHR;
+};
+#endif /* !__ASSEMBLER__ */
+
+/* WDT register fields */
 #define EWDKEYEN		BIT(5)
```
Comment (brockus-zephyr) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:1175]: all of these bit defines are great except there is no correlation to the field it belongs to. I like the field name to be prefixed by the field it belongs to but that does lead to longer names... at the very least, please add a comment to show which register each group of defines belongs with

File: drivers/watchdog/wdt_ite_it8xxx2.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+		wdt_it8xxx2_disable(dev);
+	}
+
+	/* unlock access to watchdog registers */
+	inst->ETWCFG = 0x00;
+	/* set WDT and timer1 to use 1.024kHz clock */
+	inst->ET1PSR = 0x01;
+	/* set WDT key match enabled and WDT clock to use ET1PSR */
+	inst->ETWCFG = 0x30;
+	/* watchdog can be stopped */
```
Comment (brockus-zephyr) [inline drivers/watchdog/wdt_ite_it8xxx2.c:241]: defines for the above three magic numbers would make the code more readable.

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 1 above)
Comment (RuibinChang) [reply soc/riscv/riscv-ite/common/chip_chipregs.h]: Done

File: drivers/watchdog/wdt_ite_it8xxx2.c (hunk 2 above)
Comment (RuibinChang) [reply drivers/watchdog/wdt_ite_it8xxx2.c:249]: Done

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 3]
```diff
@@ -1116,30 +1116,81 @@
...
+#define IT8XXX2_WDT_EWDSCEN		BIT(5)
+#define IT8XXX2_WDT_EWDSCMS		BIT(4)
+#define IT8XXX2_WDT_ET2TC		BIT(3)
+#define IT8XXX2_WDT_ET2RST		BIT(2)
+#define IT8XXX2_WDT_ET1TC		BIT(1)
+#define IT8XXX2_WDT_ET1RST		BIT(0)
+
 #define ET3CTRL			ECREG(EC_REG_BASE_ADDR + 0x1F10)
 #define ET3PSR			ECREG(EC_REG_BASE_ADDR + 0x1F11)
 #define ET3CNTLLR		ECREG(EC_REG_BASE_ADDR + 0x1F14)
```
Comment (katsuster) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:1196]: Do you have special reason for define these 3 symbols using macro?
Or simply forget to convert to struct?

File: dts/riscv/it8xxx2.dtsi [hunk 4]
```diff
@@ -240,7 +240,16 @@
...
-		timer: timer@f01f00 {
+
+		twd0: watchdog@f01f00 {
+			compatible = "ite,it8xxx2-watchdog";
+			reg = <0x00f01f00 0x0062>;
+			label = "TWD_0";
+			interrupts = <30 IRQ_TYPE_EDGE_RISING>;
+			interrupt-parent = <&intc>;
+		};
+
```
Comment (katsuster) [inline dts/riscv/it8xxx2.dtsi:251]: - twd means "timer for watchdog"? or typo of wdt?
- It seems this register space is overlapped on next node timer@f01f10
  reg = <0x00f01f00 0x0010> is correct?

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 3 above)
Comment (RuibinChang) [reply soc/riscv/riscv-ite/common/chip_chipregs.h:1196]: ET3CTRL, ET3PSR, and ET3CNTLLR these registers are used in timer driver. I will do another PR to re-work timer driver that including convert these macro to struct.

File: dts/riscv/it8xxx2.dtsi (hunk 4 above)
Comment (RuibinChang) [reply dts/riscv/it8xxx2.dtsi:251]: Yes, it means "timer for watchdog".
Yes, reg address 0x00f01f00 of timer node is correct and we currently get address by DT_REG_ADDR_BY_IDX now . Once macro convert to struct, this address will also change to 0x00f01f10 in timer re-work PR.

File: drivers/watchdog/Kconfig.it8xxx2 [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config WDT_ITE_IT8XXX2
+	bool "ITE it8xxx2 Watchdog Timer (WDT) driver"
+	depends on (SOC_IT8XXX2)
+	default y
+	help
+	  Enable WDT driver for ITE it8xxx2.
+
+config PLATFORM_EC_WATCHDOG_WARNING_LEADING_TIME_MS
```
Comment (keith-zephyr) [inline drivers/watchdog/Kconfig.it8xxx2:11]: I suggest changing this config name to WDT_ITE_WARNING_LEADING_TIME_MS.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: dts/riscv/it8xxx2.dtsi (hunk 4 above)
Comment (katsuster) [reply dts/riscv/it8xxx2.dtsi:251]: I understand, thanks!

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 3 above)
Comment (katsuster) [reply soc/riscv/riscv-ite/common/chip_chipregs.h:1196]: I understand, thanks!

**Review by katsuster — APPROVED** [review]

File: drivers/watchdog/Kconfig.it8xxx2 (hunk 5 above)
Comment (RuibinChang) [reply drivers/watchdog/Kconfig.it8xxx2]: Done

**Review by sjg20 — CHANGES_REQUESTED** [review]

File: drivers/watchdog/Kconfig.it8xxx2 [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2021 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config WDT_ITE_IT8XXX2
+	bool "ITE it8xxx2 Watchdog Timer (WDT) driver"
+	depends on (SOC_IT8XXX2)
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/Kconfig.it8xxx2:6]: Done

File: drivers/watchdog/Kconfig.it8xxx2 [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+	default y
+	help
+	  Enable WDT driver for ITE it8xxx2.
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/Kconfig.it8xxx2]: Done

File: drivers/watchdog/Kconfig.it8xxx2 (hunk 5 above)
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/Kconfig.it8xxx2]: As Keith suggestion, I had changed to WDT_ITE_WARNING_LEADING_TIME_MS.
So does it need to change to WDT_ITE_IT8XXX2_PERIOD?

File: drivers/watchdog/wdt_ite_it8xxx2.c [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	 * bit0 = 1: lock watchdog and timer1 config register
+	 */
+	inst->ETWCFG = (EWDKEYEN | EWDSRC | LEWDCNTL | LET1PS | LETWCFG);
+
+	LOG_DBG("WDT Setup and enabled");
+
+	return 0;
+}
+
+static int wdt_it8xxx2_feed(const struct device *dev, int channel_id)
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/wdt_ite_it8xxx2.c:145]: Done

File: drivers/watchdog/wdt_ite_it8xxx2.c [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+/* device config */
+struct wdt_it8xxx2_config {
+	/* wdt register base address */
+	uintptr_t base;
+};
+
+/* driver data */
+struct wdt_it8xxx2_data {
+	/* timeout callback used to handle watchdog event */
+	wdt_callback_t cb;
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/wdt_ite_it8xxx2.c]: Done

File: drivers/watchdog/wdt_ite_it8xxx2.c [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+#include <soc.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+LOG_MODULE_REGISTER(wdt_ite_it8xxx2);
+
+#define IT8XXX2_WATCHDOG_MAGIC_WORD			0x5c
+#define WARNING_TIMER_PERIOD_MS_TO_1024HZ_COUNT(ms)	((ms) * 1024 / 1000)
+
+/* enter critical period or not. */
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/watchdog/wdt_ite_it8xxx2.c]: Done

**Review by brockus-zephyr — APPROVED** [review]

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 11]
```diff
@@ -1760,22 +1811,6 @@
...
-#define IT83XX_ETWD_ET1CNTLLR        ECREG(IT83XX_ETWD_BASE + 0x04)
-#define IT83XX_ETWD_ETWCTRL          ECREG(IT83XX_ETWD_BASE + 0x05)
-#define IT83XX_ETWD_EWDCNTLLR        ECREG(IT83XX_ETWD_BASE + 0x06)
-#define IT83XX_ETWD_EWDKEYR          ECREG(IT83XX_ETWD_BASE + 0x07)
-#define IT83XX_ETWD_EWDCNTLHR        ECREG(IT83XX_ETWD_BASE + 0x09)
-#define IT83XX_ETWD_ETXCTRL(n)       ECREG(IT83XX_ETWD_BASE + 0x10 + (n << 3))
-#define IT83XX_ETWD_ETXPSR(n)        ECREG(IT83XX_ETWD_BASE + 0x11 + (n << 3))
-#define IT83XX_ETWD_ETXCNTLR(n)      ECREG_u32(IT83XX_ETWD_BASE + 0x14 + (n << 3))
-#define IT83XX_ETWD_ETXCNTOR(n)      ECREG_u32(IT83XX_ETWD_BASE + 0x48 + (n << 2))
-
```
Comment (RuibinChang) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:1778]: Delete these unused definition.

**Comment by RuibinChang** [comment]
Any more comment here?
There is a relation chain depending on this change on
https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2914711

**Review by keith-zephyr — APPROVED** [review]

**Comment by RuibinChang** [comment]
Could anyone help me to merge this changes?

File: dts/riscv/it8xxx2.dtsi [hunk 12]
```diff
@@ -240,7 +240,16 @@
...
+		timer: timer@f01f10 {
```
Comment (sjg20) [inline dts/riscv/it8xxx2.dtsi:252]: If you update this, please also update the 'reg' property below

File: drivers/watchdog/wdt_ite_it8xxx2.c [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+#include <logging/log.h>
+#include <drivers/watchdog.h>
+#include <soc.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+LOG_MODULE_REGISTER(wdt_ite_it8xxx2);
+
+#define IT8XXX2_WATCHDOG_MAGIC_WORD			0x5c
```
Comment (sjg20) [inline drivers/watchdog/wdt_ite_it8xxx2.c:16]: This is a byte, not a word (neither 16- or 32-bit)

**Review by sjg20 — CHANGES_REQUESTED** [review]

File: dts/riscv/it8xxx2.dtsi (hunk 12 above)
Comment (RuibinChang) [reply dts/riscv/it8xxx2.dtsi:252]: Yes, I will.

File: drivers/watchdog/wdt_ite_it8xxx2.c (hunk 13 above)
Comment (RuibinChang) [reply drivers/watchdog/wdt_ite_it8xxx2.c:16]: Done

**Comment by brockus-zephyr** [comment]
Can someone with merge authority please land this?

**Review by sjg20 — APPROVED** [review]

**Pull request merged**